src/language/pattern-types.md

```diff
@@ -0,0 +1,333 @@
+---
+title: Pattern types
+description: Pattern type reference in Dart.
+---
+
+## Cast
+
+`foo as String`
+
+Cast patterns let you insert a [type cast][] as a pattern is
+[destructuring][destructure] values:
+
+```dart
+(num, Object) record = (1, "s");
+var (i as int, s as String) = record;
+```
+
+Cast patterns will [throw][] if the value doesn't have the stated type.
+Like the [null-assert pattern](#null-assert), this lets you forcibly assert the
+expected type of some destructured value.
+
+## Constant	
+
+`123, null, 'string', math.pi, SomeClass.constant, const Thing(1, 2), const (1 + 2)`
+
+Constant patterns allow simple literals and references to named constants to be
+used directly as patterns. They match when the value is equal to the constant.
+
+Some valid constant expressions may not be valid constant patterns. In that case,
+to distinguish a constant expression from a constant pattern, prefix the expression
+with `const` and wrap it in parenthesis so it is unambiguosly an expression and
+not a constant pattern:
```

Comment:
How about:

"Number (`123`, `45.56`), boolean (`true`), and string (`'hi'`) literals, and named constants (`someConstant`, `math.pi`, `double.infinity`) are directly usable as constant patterns. Constant constructors (`const Point(0, 0)`) and constant collection literals (`const []`, `const {1, 2}`) can also be used directly as patterns. More complex constant expressions must be parenthesized and prefixed with `const` (`const (1 + 2)`)."

(Or maybe pull those examples out into code blocks.)